protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,294 @@
+                    .push_back(ToSwarm::ExternalAddrConfirmed(reachable_addr));
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(
+                Err(dial_request::Error::UnableToConnectOnSelectedAddress { addr: Some(addr) })
+            ))
+            | Either::Left(dial_request::ToBehaviour::TestCompleted(
+                Err(dial_request::Error::FailureDuringDialBack { addr: Some(addr) })
+            )) => {
+                self.pending_events
+                    .push_back(ToSwarm::ExternalAddrExpired(addr));
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(Err(err))) => {
+                tracing::debug!("Test failed: {:?}", err);
+            }
+        }
+    }
+
+    fn poll(
+        &mut self,
+        _cx: &mut Context<'_>,
+    ) -> Poll<ToSwarm<Self::ToSwarm, <Handler as ConnectionHandler>::FromBehaviour>> {
+        let pending_event = self.poll_pending_events();
+        if pending_event.is_ready() {
+            return pending_event;
+        }
+        if self.ticker.ready() && !self.known_servers.is_empty() {
+            let mut entries = self.address_candidates.drain().collect::<Vec<_>>();
+            entries.sort_unstable_by_key(|(_, count)| *count);
+            let addrs = entries
+                .into_iter()
+                .rev()
+                .map(|(addr, _)| addr)
+                .take(self.config.max_addrs_count)
+                .collect::<Vec<_>>();
+            let peers = if self.known_servers.len() < self.config.test_server_count {
+                self.known_servers.clone()
+            } else {
+                self.known_servers
+                    .choose_multiple(&mut self.rng, self.config.test_server_count)
+                    .copied()
+                    .collect()
+            };
+            for peer in peers {
+                let nonce = self.rng.gen();
+                let req = DialRequest {
+                    nonce,
+                    addrs: addrs.clone(),
+                };
+                self.pending_nonces.insert(nonce);
+                self.submit_req_for_peer(peer, req);
+            }
+            let pending_event = self.poll_pending_events();
+            if pending_event.is_ready() {
+                return pending_event;
+            }
+        }
+        Poll::Pending
+    }
+}
+
+impl<R> Behaviour<R>
+where
+    R: RngCore + 'static,
+{
+    fn submit_req_for_peer(&mut self, peer: PeerId, req: DialRequest) {
+        if let Some(conn_id) = self.peers_to_handlers.get(&peer) {
+            self.pending_events.push_back(ToSwarm::NotifyHandler {
+                peer_id: peer,
+                handler: NotifyHandler::One(*conn_id),
+                event: Either::Left(dial_request::FromBehaviour::PerformRequest(req)),
+            });
+        } else {
+            tracing::debug!(
+                "There should be a connection to {:?}, but there isn't",
+                peer
+            );
```

Comment:
If we store `(PeerId, ConnectionId)` for each discovered autonat server, this error case should go away :partying_face:

Reply:
yes you are right. I will remove it.